core/src/main/scala/org/apache/spark/deploy/security/README.md

```diff
@@ -0,0 +1,249 @@
+on an external renewal service (like YARN). That way, Spark's renewal feature can be used with
+resource managers that are not DT-aware, such as Mesos or Kubernetes, as long as the application
+has access to a keytab.
+
+
+## DTs and Proxy Users
+
+"Proxy users" is Hadoop-speak for impersonation. It allows user A to impersonate user B when
+connecting to a service, if that service allows it.
+
+Spark allows impersonation when submitting applications, so that the whole application runs as
+user B in the above example.
+
+Spark does not allow token renewal when impersonation is on. Impersonation was added in Spark
+as a means for services (like Hive or Oozie) to start Spark applications on behalf of users.
+That means that those services would provide the Spark launcher code with privileged credentials
+and, potentially, user code that will run when the application starts. The user code is not
+necessarily under control of the service.
+
+In that situation, the service credentials should never be made available to the Spark application,
+since that would be tantamount to giving your service credentials to unprivileged users.
+
+The above also implies that running impersonated applications in client mode can be a security
+concern, since arbitrary user code would have access to the same local content as the privileged
+user. But unlike token renewal, Spark does not prevent that configuration from running.
+
+When impersonating, the Spark launcher will create DTs for the "proxy" user. In the example
+used above, that means that when code authenticates to a service using the DTs, the authenticated
+user will be "B", not "A".
+
+Note that "proxy user" is a very Hadoop-specific concept. It does not apply to OS users (which
+is why the client-mode case is an issue) and to services that do not authenticate using Hadoop's
+`UserGroupInformation` system. It is generally used in the context of YARN - since an application
+submitted as a proxy user will run as that particular user in the YARN cluster, obeying any
+Hadoop-to-local-OS-user mapping configured for the service. But the overall support should work
+for connecting to other services even when YARN is not being used.
+
+Also, if writing a new DT provider in Spark, be aware that providers need to explicitly handle
+impersonation. If a service does not support impersonation, the provider should either error out or
+not generate tokens, depending on what makes more sense in the context.
+
+
+## Externally Generated DTs
+
+Spark uses the `UserGroupInformation` API to manage the Hadoop credentials. That means that Spark
+inherits the feature of loading DTs automatically from a file. The Hadoop classes will load the
+token cache pointed at by the `HADOOP_TOKEN_FILE_LOCATION` environment variable, when it's defined.
+
+In this situation, Spark will not create DTs for the services that already have tokens in the
```

Comment:
@tgravescs Since not many peoples are dealing with security maybe just you and me so asking you directly. I've just double checked the code because wanted to use the mentioned feature but I have the feeling it's not working. After some digging I've not seen any significant code changes in the security area so I have the feeling it's maybe never worked. Have you ever looked after this feature? When I take a look at the current code I see that the obtained tokens are overwriting the existing ones.

Reply:
As far as I can remember this worked at one point and time, but there has been a lot of changes to the security code since I possibly tried it. I don't really deal with security anymore so don't have a setup to try.

Reply:
Good to hear at least once it was working. Do you know by heart roughly where to look for the mentioned filtering?
I though I would see some if condition not to set UGI but there is no such code.

Reply:
Sounds like the problem is that it picks them up by then overwrites?  I would look for the calls to addCredentials, I don't remember the details on that api to know if they overwrite or merge things. Seems like at one point we specifically handled merging but like I said its been a long time since I wrote that code and many people changed it since.

Reply:
The HADOOP_TOKEN_FILE_LOCATION stuff was just all handled by Hadoop for us and we inherit so might not see any code directly related to that. If its overwriting them, my guess is we are just smashing in the tokens rather then dealing with them there.

Reply:
Thx for the pointers, going to sit together with the hadoop guys and having a look at HADOOP_TOKEN_FILE_LOCATION + addCredentials in depth. AFIAR addCredentials overwrites by default but will see... Thx again!